### ISSUE TITLE
Pair type with swapped K V Generic variables is reported as Pair[Never, Never]

### ISSUE THREAD

**Issue by thomasaarholt** [astral-sh/ty#2817]
### Summary

I searched but couldn’t find this reported already. Really excited about what you’re building here.

Inside the swap method, `Pair(self.value, self.key)` is inferred as `Pair[Never, Never]`, while the call site infers the correct `Pair[int, str]`. 

Pair[Never, Never] is incorrect

[playground](https://play.ty.dev/6abd4b54-f69a-40a2-b9d3-c4131242732e)
```python
from typing import reveal_type


class Pair[K, V]:
    def __init__(self, key: K, value: V) -> None:
        self.key: K = key
        self.value: V = value

    def swap(self) -> Pair[V, K]:
        swapped_pair = Pair(self.value, self.key)
        reveal_type(swapped_pair) # Revealed type: `Pair[Never, Never]`
        return swapped_pair # Return type does not match returned value: expected `Pair[V@Pair, K@Pair]`, found `Pair[Never, Never]` (invalid-return-type) [Ln 12, Col 16]

pair = Pair("foo", 123)  # Pair[str, int]
swapped = pair.swap()  # correctly infers swapped type Pair[int, str]
```

### Version

ty 0.0.17 (8cec85718 2026-02-13)

**Comment by carljm** [comment]
Interesting, thanks for the report!

It looks like we produce mappings `K -> V` and `V -> K` here, and then finalizing those specializations recursively, we consider it to be a cycle and break that cycle, resulting in both typevars solving to `Never`.

I haven't thought deeply about it, but it feels like an issue where we are wrongly considering the `K` and `V` typevars of `self` to be the same typevar as `K` and `V` in the newly-constructed `Pair`, which is wrong -- those typevars are independent.

**Comment by sharifhsn** [comment]
The example here now reports as `Pair[K@Pair, K@Pair]` after ~~https://github.com/astral-sh/ruff/pull/24852~~  https://github.com/astral-sh/ruff/pull/24582 refactored generics and got rid of the cycle-break to `Never` in this case, which suggests that the remaining issue relates to type-var substitution, as remarked on in #3399.

**Comment by sharkdp** [comment]
> The example here now reports as `Pair[K@Pair, K@Pair]` after [astral-sh/ruff#24852](https://github.com/astral-sh/ruff/pull/24852)

is that a wrong reference? the linked PR hasn't been merged, and is not a generics refactor.

**Comment by sharifhsn** [comment]
I meant https://github.com/astral-sh/ruff/pull/24582 🤦🏽

**Comment by thomasaarholt** [comment]
This appears fixed on playground! #3984 appears fixed too!
